tests/MoneyFormatterTest.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+namespace App\Tests;
+
+use App\Services\MoneyFormatter;
+use App\Services\NumberFormatter;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+class MoneyFormatterTest extends TestCase
+{
+
+    /** @var NumberFormatter|MockObject $numberFormatter */
+    private $numberFormatter;
+
+    /**
+     * @param int $index
+     * @param float $argument
+     * @param string $returnValue
+     */
+    private function addCase(int $index, float $argument, string $returnValue)
+    {
+        $this->numberFormatter->expects($this->at($index))
+            ->method('format')
+            ->with($argument)
+            ->will($this->returnValue($returnValue));
+    }
+
+    public function testFormatCurrency()
+    {
+        $this->numberFormatter = $this->createMock(NumberFormatter::class);
+
+        $this->addCase(0, 2835779, '2.84M');
```

Comment:
Įdomus sprendimas...

Realybėje dažniau būna tiesiog būna mažiau testavimo atvejų.
`NumberFormatterTest` patikrina visus _mega_ ir _kilo_ variantus. Tai `MoneyFormatterTest` tau reikėtų koncentruotis tik su `$` ir `€` atvejams.

T.y. užtektų ir:
```php
        $this->assertEquals('2.84M €', $moneyFormatter->formatEur(2835779));
        $this->assertEquals('$211.99', $moneyFormatter->formatUsd(211.99));
```

Nes kai kodas keičiasi, tai reikia atnaujinti ir testus. Jei visi nuo visko priklauso – tai testų taisymas užimtų daug laiko. **`Unit`** testavimo esmė ir yra testuoti tik tai 1 klasę/elementą – atsiribojant nuo visų kitų klasių/priklausomybių.
Pvz. jei `NumberFormatter` klasė pakeista, tai pagauti turėtų `NumberFormatterTest`, o ne `MoneyFormatterTest`.

Praktikoje labiau norisi, kad atlikus pakeitimą sulūžti tik 1, o ne 21 testas – nes tai veikia kaip dokumentacija – žinoma, kaip tavo pakeitimas įtakos likusią sistemą.

Plačiau:
* https://en.wikipedia.org/wiki/Single_responsibility_principle
* https://medium.com/pacroy/why-most-unit-testing-is-waste-tests-dont-improve-quality-developers-do-47a8584f79ab#abdf